### ISSUE TITLE
[$200]Registration issue for Challenges requiring NDA

### ISSUE THREAD

**Issue by SathyaJayabal** [topcoder-platform/community-app#4609]
For a challenge requiring NDA (https://beta-community-app.topcoder.com/challenges/e77d97be-a576-4d62-966b-3a46b2d3f530) the NDA popup is displayed even though user (picachui) has already agreed to it previously (in V4)
The terms are not displayed completely and there is no agree button to continue.
<img width="1440" alt="Screenshot 2020-06-26 at 9 09 25 AM" src="https://user-images.githubusercontent.com/58783823/85823493-337f0e00-b79b-11ea-893d-c52324067d01.png">

**Comment by crazyk07** [comment]
Contest https://www.topcoder.com/challenges/30131350 has been created for this ticket.<br/><br/>```This is an automated message for crazyk via Topcoder X```

**Comment by crazyk07** [comment]
Contest https://www.topcoder.com/challenges/30131350 has been updated - it has been assigned to luizrrodrigues.<br/><br/>```This is an automated message for crazyk via Topcoder X```

**Comment by luizrrodrigues** [comment]
@rootelement V5 Terms missing **docusignTemplateId** to **DocuSignable** term type:

![image](https://user-images.githubusercontent.com/7029641/86114528-08e5cb80-baa1-11ea-8932-2031237f9d92.png)

swagger file:
https://github.com/topcoder-platform/terms-service/blob/624ca31a830ab448af91c96776ba5ccd91336df1/docs/swagger.yaml#L204-L205

V2 return:
![image](https://user-images.githubusercontent.com/7029641/86114551-10a57000-baa1-11ea-974d-95e18cda82ae.png)

**Comment by rootelement** [comment]
@luizrrodrigues i _think_ this was due to the docusignable template id being wrong in the config. Deploying now and testing.

**Comment by rootelement** [comment]
Looks like it works in dev:
![Screen Shot 2020-06-30 at 10 41 45 AM](https://user-images.githubusercontent.com/5400156/86139951-53297580-babe-11ea-9b7d-2e7f42124076.png)

**Comment by luizrrodrigues** [comment]
@rootelement This returning well now,  but now jump in a new issue:

1) When try to request `/v5/terms/docusignViewURL` receiving this error:
`Login to DocuSign server failed.`

![image](https://user-images.githubusercontent.com/7029641/86184250-2a2ed200-bb0a-11ea-8a60-54eb0840a1f0.png)

-----

2) Looks like not need pass **returnUrl** anymore, but instead need pass **tabs**, can you help with this?

**Comment by rootelement** [comment]
Status update here. Everything works except for the Docusign Callback. Working on that now.

**Comment by rootelement** [comment]
This is unblocked from the API side. @luizrrodrigues is working on cleaning up the UI.

**Comment by sushilshinde** [comment]
@luizrrodrigues waiting for PR

**Comment by luizrrodrigues** [comment]
@sushilshinde I'm waiting for API fix.

@rootelement Working on this.

**Comment by luizrrodrigues** [comment]
@rootelement API returning error `Requesting Signature via template failed.`

![image](https://user-images.githubusercontent.com/7029641/86603640-a138f000-bf7a-11ea-896d-657cd30ccafd.png)

**Comment by rootelement** [comment]
I just tried that very same POST with my jcori dev account, which has these roles:
```
{
  "roles": [
    "copilot",
    "administrator",
    "Connect Admin",
    "Topcoder User"
  ],
  "iss": "https://api.topcoder-dev.com",
  "handle": "jcori",
  "exp": 1594221019,
  "userId": "40029484",
  "iat": 1594220419,
  "email": "jcori+dev@topcoder.com",
  "jti": "89df045b-89a7-405c-bc92-61cce1ac6cf4"
}
```

And it worked fine. I got the URL by POSTing from Postman, opened the url returned in the browser, signed the document, and was given the `{'message': 'success'}` return.

I was even notified of the signature because I own the demo instance:
<img width="1304" alt="Screen Shot 2020-07-08 at 11 04 27 AM" src="https://user-images.githubusercontent.com/5400156/86935469-decc8300-c10a-11ea-876a-13f475022295.png">

Can you please post your steps @luizrrodrigues ?

**Comment by mtwomey** [comment]
@luizrrodrigues / @sushilshinde Further notes:

**Test Token**
```
{
  "roles": [
    "Topcoder User"
  ],
  "iss": "https://api.topcoder-dev.com",
  "handle": "mtwomey40",
  "exp": 1594222439,
  "userId": "88773500",
  "iat": 1594221839,
  "email": "mtwomey40@beakstar.com",
  "jti": "1ae04bf5-4c87-4ab2-97ad-ef2af5920ab9"
}
```

**Curl**
```
curl --location --request POST 'https://api.topcoder-dev.com/v5/terms/docusignViewURL' \
--header 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJyb2xlcyI6WyJUb3Bjb2RlciBVc2VyIl0sImlzcyI6Imh0dHBzOi8vYXBpLnRvcGNvZGVyLWRldi5jb20iLCJoYW5kbGUiOiJtdHdvbWV5NDAiLCJleHAiOjE1OTQyMjI0MzksInVzZXJJZCI6Ijg4NzczNTAwIiwiaWF0IjoxNTk0MjIxODM5LCJlbWFpbCI6Im10d29tZXk0MEBiZWFrc3Rhci5jb20iLCJqdGkiOiIxYWUwNGJmNS00Yzg3LTRhYjItOTdhZC1lZjJhZjU5MjBhYjkifQ.DMNtD3LHIqekIM6J4_zv2CrpC-ho9gRFTGfikvpRbZE' \
--header 'Content-Type: application/json' \
--header 'Cookie: tcsso=40029484|a1dbc7405f7186d4df353764b5bdeaf5a9d954a43d09c4ba4baad730f66818; tcjwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuaWNrbmFtZSI6Impjb3JpIiwiZW1haWwiOiJqY29yaStkZXZAdG9wY29kZXIuY29tIiwibmFtZSI6Impjb3JpK2RldkB0b3Bjb2Rlci5jb20iLCJwaWN0dXJlIjoiaHR0cHM6Ly9zLmdyYXZhdGFyLmNvbS9hdmF0YXIvZTk0ODU2ZDAxMThmNjkwYjY2MWRjN2NjMWJjODFkOGY_cz00ODAmcj1wZyZkPWh0dHBzJTNBJTJGJTJGY2RuLmF1dGgwLmNvbSUyRmF2YXRhcnMlMkZqYy5wbmciLCJyb2xlcyI6WyJ1c2VyIl0sImFwcF9tZXRhZGF0YSI6eyJyb2xlcyI6WyJ1c2VyIl0sImF1dGhvcml6YXRpb24iOnsiZ3JvdXBzIjpbXX19LCJhdXRob3JpemF0aW9uIjp7Imdyb3VwcyI6W119LCJncm91cHMiOltdLCJjbGllbnRJRCI6IkpGRG83SE1rZjBxMkNrVkZIb2p5M3pIV2FmemlwcmhUIiwidXBkYXRlZF9hdCI6IjIwMjAtMDYtMTVUMjA6NDc6MzEuNzEzWiIsInVzZXJfaWQiOiJhdXRoMHw0MDAyOTQ4NCIsImlkZW50aXRpZXMiOlt7InVzZXJfaWQiOiI0MDAyOTQ4NCIsInByb3ZpZGVyIjoiYXV0aDAiLCJjb25uZWN0aW9uIjoiVEMtVXNlci1EYXRhYmFzZSIsImlzU29jaWFsIjpmYWxzZX1dLCJjcmVhdGVkX2F0IjoiMjAxNy0xMS0xMFQxOTowNzoxNi4xNDlaIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsImlzcyI6Imh0dHBzOi8vdG9wY29kZXItZGV2LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw0MDAyOTQ4NCIsImF1ZCI6IkpGRG83SE1rZjBxMkNrVkZIb2p5M3pIV2FmemlwcmhUIiwiaWF0IjoxNTkyMjU0MDUzLCJleHAiOjE5NTIyNTQwNTN9.ZcLFfo2y--aJLRtc9NUaY97H-bsH73jJF0rOZJCNVwM' \
--data-raw '{ 
	"templateId": "7893116a-bf76-475a-b92e-3931c00902a1"
}'
```

**Request Response (200)**
```
{
    "recipientViewUrl": "https://demo.docusign.net/Signing/MTRedeem/v1/1a5b37c6-bcc5-48eb-a73c-8b0ce9e37560?slt=eyJ0eXAiOiJNVCIsImFsZyI6IlJTMjU2Iiwia2lkIjoiNjgxODVmZjEtNGU1MS00Y2U5LWFmMWMtNjg5ODEyMjAzMzE3In0.AQUAAAABAAMABwAAnT8hUyPYSAgAAPsP1FMj2EgYAAEAAAAAAAAAIQBdAwAAeyJUb2tlbklkIjoiYzMyZjkwZDAtOGYyMy00MzZhLWFlNTctN2Q1MDgxZGRkZjAxIiwiU3ViamVjdCI6bnVsbCwiU3ViamVjdFR5cGUiOm51bGwsIkV4cGlyYXRpb24iOiIyMDIwLTA3LTA4VDE1OjMwOjIyKzAwOjAwIiwiSXNzdWVkQXQiOiIyMDIwLTA3LTA4VDE1OjI1OjIyLjE3NDc4OTQrMDA6MDAiLCJSZXNvdXJjZUlkIjoiNzc1ZTViZDItZTg1MC00NjI0LThiYzItODFlZWJjYjI0ODBiIiwiTGFiZWwiOm51bGwsIlNpdGVJZCI6bnVsbCwiUmVzb3VyY2VzIjoie1wiRW52ZWxvcGVJZFwiOlwiNzc1ZTViZDItZTg1MC00NjI0LThiYzItODFlZWJjYjI0ODBiXCIsXCJBY3RvclVzZXJJZFwiOlwiNmJlZDVkYTctNDA3OC00N2Y3LTk1OGUtM2FmOTA4OWQ0ZDQ1XCIsXCJSZWNpcGllbnRJZFwiOlwiNmExNjBjZDMtODczMC00N2M2LWIxMDYtM2I2OTZkYjViZjRjXCIsXCJGYWtlUXVlcnlTdHJpbmdcIjpcInQ9MGM0ZWIyZjItMjRkZC00YzY2LWExZGMtNGQwMmQyZTIyYmY4XCJ9IiwiT0F1dGhTdGF0ZSI6bnVsbCwiVG9rZW5UeXBlIjoxLCJBdWRpZW5jZSI6IjI1ZTA5Mzk4LTAzNDQtNDkwYy04ZTUzLTNhYjJjYTU2MjdiZiIsIlNjb3BlcyI6bnVsbCwiUmVkaXJlY3RVcmkiOiJodHRwczovL2RlbW8uZG9jdXNpZ24ubmV0L1NpZ25pbmcvU3RhcnRJblNlc3Npb24uYXNweCIsIkhhc2hBbGdvcml0aG0iOjAsIkhhc2hTYWx0IjpudWxsLCJIYXNoUm91bmRzIjowLCJUb2tlblNlY3JldEhhc2giOm51bGwsIlRva2VuU3RhdHVzIjowLCJFeHRlcm5hbENsYWltc1JlcXVlc3RlZCI6bnVsbCwiVHJhbnNhY3Rpb25JZCI6bnVsbCwiVHJhbnNhY3Rpb25FdmVudENhbGxiYWNrVXJsIjpudWxsLCJJc1NpbmdsZVVzZSI6ZmFsc2V9.1_W0UAVwWntFN0AHc1BxgfKzmriN_SvehG6Li-ff-XqW33HrrFRW6jv0kffTWYt5E4GeJq-XG8Gjztg0lVLsYzBIVe70b4M2libYNSVGwIyv_VnQYs9XoEIB-MYsGXZsjoA1WsXMs9yY0NrAKiF6p3H_ZvyHeUeVyKuloF0kKbmoOg47TXS8BX-XkyZJsXy4lViaFPmbSzCLnpsieuC--xtUWPH5h2L_rl7cLiMG3xAhvdIIXYEGKRtyHc3JPNWN18zZP8V2JlYWulJ2S51skrax-FOfyqeuod7BzBjBWoAOUg3l-wkj_Dk7pbrE0uduHYv_Ofb1ADN6ga68hiiLfg",
    "envelopeId": "775e5bd2-e850-4624-8bc2-81eebcb2480b"
}
```

**Comment by luizrrodrigues** [comment]
Let me check this again and back with the results.

**Comment by luizrrodrigues** [comment]
@mtwomey @rootelement @sushilshinde 

API looks like working well now, I'll push PR shortly.

**Comment by luizrrodrigues** [comment]
@rootelement When backend receives a return from DocuSign and set this as completed:
https://github.com/topcoder-platform/terms-service/blob/36c2b4d8bdecf9470961807ab5b2088e10d77348/src/services/DocusignService.js#L86

The backend should now return this terms `agree: true` right?

**Comment by luizrrodrigues** [comment]
@rootelement 
Return looks good to **dan_developer** user, but getting error to **luizrrodrigues** user:

```
{
  "roles": [
    "copilot",
    "Topcoder User"
  ],
  "iss": "https://api.topcoder-dev.com",
  "handle": "luizrrodrigues",
  "exp": 1594350188,
  "userId": "22827870",
  "iat": 1594349588,
  "email": "email@domain.com.z",
  "jti": "e0aa2a22-3bbf-4923-a56f-bc2672802217"
}
```

Error when request `https://api.topcoder-dev.com/v5/terms/docusignViewURL`

```
"message": "Requesting Signature via template failed."
```

**Comment by mtwomey** [comment]
Hoping to have the first draft of API updates tomorrow (7/13/2020). (Thomas K is working this)

**Comment by mtwomey** [comment]
Thomas still working.

**Comment by sushilshinde** [comment]
@ThomasKranitsas updates on this?

**Comment by rootelement** [comment]
@sushilshinde I'm working with Thomas. He has a to-do list for the start of his day. We should have this functioning in dev today. I will update when I get back to my desk.

@luizrrodrigues There will be a change in how you implement this. Couple things:
First, we removed the stupid `noauth` thing. If you pass JWT, it uses it, if you don't, it doesn't. M2M and Admin JWT don't need to pass `?memberId`, but if they do, it'll say if they signed it or not. I believe Thomas has updated the swagger, but will confirm that.

Here's an updated summary of the steps from that conversation we had with Matt, amended for how it will work now. We should have this done sometime today (July 16th)

1. FE calls terms API for user to see if agreed is true, if not:
2. If term is DocuSign type, FE call `POST /terms/docusignViewURL` to get DocuSign URL This call will accept a body containing {templateId, tabs (not required), returnUrl} Return URL can be whatever you want it to be on your end. This returns the docusign url
3. FE sends the user to the Docusign url.
4. User signs the Docusign document.
5. Docusign calls an authenticated backend webhook on our side (/v5/terms-receiver/webhook) to update the terms signature for the envelopeId
6. When Docusign redirects the user to the returnUrl you passed, you call the same GET as Step 1 again to see if the terms are agreed. You _might_ have to poll that for a couple seconds with a timeout. This will be the part we have to test.

Again, I tell you all of this to tell you what _will be_ there later today. It's not deployed in dev yet. But if you want to start working on implementing this pattern in anticipation of it being there, go ahead.

I've updated the $$ of this challenge since it has lingered and we've added rework.

**Comment by luizrrodrigues** [comment]
@rootelement cool, looks good, I'll start with updates tonight (my time)

And thanks for raise the payment, I appreciated that.

**Comment by rootelement** [comment]
This is confirmed working.

Get Terms with User JWT
```
curl --location --request GET 'https://api.topcoder-dev.com/v5/terms/77f558c1-56fb-427c-b974-61ea0a060ca7' \
--header 'Authorization: Bearer <<USER JWT>>'
```
response:
```json
{
    "id": "77f558c1-56fb-427c-b974-61ea0a060ca7",
    "title": "Competition Non-Disclosure Agreement",
    "url": "",
    "text": "docusign NDA",
    "typeId": 5,
    "legacyId": 21343,
    "agreed": false,
    "docusignTemplateId": "7893116a-bf76-475a-b92e-3931c00902a1",
    "agreeabilityType": "DocuSignable"
}
```

Get Docusign Envelope URL
```
curl --location --request POST 'https://api.topcoder-dev.com/v5/terms/docusignViewURL' \
--header 'Authorization: Bearer <<USER JWT>>' \
--header 'Content-Type: application/json' \
--data-raw '{
    "templateId": "7893116a-bf76-475a-b92e-3931c00902a1",
    "returnUrl": "https://www.topcoder.com"
}'
```

Get Status for User again (first request), response should now be:
```json
{
    "id": "77f558c1-56fb-427c-b974-61ea0a060ca7",
    "title": "Competition Non-Disclosure Agreement",
    "url": "",
    "text": "docusign NDA",
    "typeId": 5,
    "legacyId": 21343,
    "agreed": true,
    "docusignTemplateId": "7893116a-bf76-475a-b92e-3931c00902a1",
    "agreeabilityType": "DocuSignable"
}
```

The webhook can take 4 or 5 seconds to call back. You'll have to show an animation or something...

**Comment by rootelement** [comment]
@luizrrodrigues Any update here?

**Comment by luizrrodrigues** [comment]
@rootelement sorry delay, I wasn't to start work yesterday, I'll jump on this in 1h.

**Comment by luizrrodrigues** [comment]
@rootelement Cool, this working well and just previous PR will works because currently flow handle with changes.

PR add a check if the return is **viewing_completed** in case the user already signed but for some reason closed the window before backend receives callback. 

-----

But have issue with user **luizrrodrigues** (tests above is done with **dan_devleoper** user)

Getting error: `Requesting Signature via template failed.` when call `https://api.topcoder-dev.com/v5/terms/docusignViewURL`

**Comment by rootelement** [comment]
I tested this with a brand new user `jcori30` and it worked great! I'm not sure what's up with your user. You might be in some sort of limbo on it. I will try clearing it out later tonight and we can try again with your user.

**Comment by luizrrodrigues** [comment]
@rootelement Yep, looks like working well with other users, I think `dan_developer` was the same issue before but looks like fixed now.

**Comment by sushilshinde** [comment]
https://github.com/topcoder-platform/community-app/pull/4640

**Comment by SathyaJayabal** [comment]
@luizrrodrigues @sushilshinde , when I try to register to an  NDA challenge, I get displayed the completed NDA document (as I have already signed it long back) with option to close. After closing the popup, I am not registered. I still see the blue register button. Clicking on register repeats the above. So I am not able to proceed and register to the challenge.

user:picachui
https://beta-community-app.topcoder.com/challenges/04e3e195-835a-4b8d-850e-39dbb99dc3cf?tab=registrants
https://beta-community-app.topcoder.com/challenges/0c8061f0-076f-47d3-a901-7cf13f99500c

Expected Behaviour: When the user has aleady signed the NDA, They should be able to go ahead and register without display of the NDA form.

**Comment by luizrrodrigues** [comment]
@SathyaJayabal Okay, I have same issue with my user in beta.

@rootelement Backend support **viewing_complete** event return from DocuSign?

- **viewing_complete** is about see already signed

**Comment by SathyaJayabal** [comment]
@luizrrodrigues @rootelement , also for user whop has not signed the NDA, the document keeps on loading for a few mins and then shows a blank page.
<img width="1440" alt="Screenshot 2020-07-21 at 4 41 48 PM" src="https://user-images.githubusercontent.com/58783823/88048544-232b3a80-cb71-11ea-90c6-3cd3607aa4ca.png">
https://drive.google.com/file/d/1s7w8euHOGHcUbn0qma-FeIB5sR0I93VK/view?usp=sharing

**Comment by luizrrodrigues** [comment]
@SathyaJayabal can you provide the user/pass to test n beta?

**Comment by SathyaJayabal** [comment]
user: picachuitest1/appirio123

**Comment by sushilshinde** [comment]
please share creds on slack for beta

**Comment by luizrrodrigues** [comment]
@SathyaJayabal This happen after sign the DocuSign or before?

I mean, you able to see the DocuSign frame?

**Comment by SathyaJayabal** [comment]
shared creds on slack @luizrrodrigues

**Comment by SathyaJayabal** [comment]
@luizrrodrigues , for a user who has not signed the NDA, the NDA document is displayed as expected. When the user signs and clicks on finish, the document loads again and the user is asked to sign again. This repeats and the the user is not able to proceed.

Please see linked video for reference. user:CustomerUser (password shared on slack)
https://drive.google.com/file/d/1qZHPP-rU9qurIx4CTITryiVnH6YS6SA5/view?usp=sharing

**Comment by luizrrodrigues** [comment]
@SathyaJayabal Thanks

@rootelement FE receiving `signing_complete` from **DocuSign**, is BE getting an error? Looks like not updating the terms assigned.

**Comment by rootelement** [comment]
Backend receiver in prod was having an error. I just was able to complete NDA for my user in prod just now:
<img width="1265" alt="Screen Shot 2020-07-23 at 4 12 31 PM" src="https://user-images.githubusercontent.com/5400156/88334202-6d422680-ccff-11ea-93b8-2c6a77c57556.png">

Please test again.

**Comment by sushilshinde** [comment]
@luizrrodrigues can you confirm and PR if there are changes?

**Comment by luizrrodrigues** [comment]
@rootelement Cool, looks good now.

@sushilshinde No changes needed, this was fixed in BE side.

**Comment by SathyaJayabal** [comment]
@luizrrodrigues , not able to register for user who has already signed the NDA. Please try with your own handle and check please.

See linked video for reference.
https://drive.google.com/file/d/1xqBBNRuNRkqh71zFSsm4Vq0U5qaGXk5A/view?usp=sharing

A user who has not yet signed the NDA is able to sign and register to a challenge
attached video
https://drive.google.com/file/d/1X3yocguu1UT0mFHGKZY9r2ceHdF5_qjH/view?usp=sharing

**Comment by SathyaJayabal** [comment]
cc @sushilshinde

**Comment by luizrrodrigues** [comment]
@rootelement Backend support viewing_complete event return from DocuSign?

viewing_complete is about to see already signed.

**Comment by luizrrodrigues** [comment]
@rootelement I see same issue with my user, id: 22827870

Challenge: https://beta-community-app.topcoder.com/challenges/658ae590-b249-4db0-bad4-b2ca86f31eda

**Comment by SathyaJayabal** [comment]
@rootelement , See the same issue with another user drasticdpk. This behaviour may be common to all users who have already signed the NDA.

**Comment by SathyaJayabal** [comment]
Verified for following users who have already signed NDA in V4 (callmekatootie, drasticdpk and kiril.kartunov ) and for users users who signed NDA in v5(picachuitest1) and users who have not yet signed NDA(picachuitest2).

Works fine for all cases.

**Comment by crazyk07** [comment]
Payment task has been updated: https://software.topcoder.com/review/actions/ViewProjectDetails?pid=30131350<br/><br/>```This is an automated message for crazyk via Topcoder X```

**Comment by rootelement** [comment]
Thank you @luizrrodrigues for your patience in testing this. And @SathyaJayabal for your awesome diligence on providing clear feedback.

**Comment by luizrrodrigues** [comment]
@rootelement No problem :)

**Comment by lakshmiathreya** [comment]
Code freeze Re-testing: Verified with member who has accepted NDA in v4 earlier (lakshmiathreya) and member who hadnt accepted earlier - accepted in v5 now (lakshmiaCopilot).
